activerecord/lib/active_record/autosave_association.rb

```diff
@@ -200,13 +200,13 @@ def add_autosave_association_callbacks(reflection)
             after_create save_method
             after_update save_method
           else
-            define_non_cyclic_method(save_method) { save_belongs_to_association(reflection) }
+            define_non_cyclic_method(save_method) { throw(:abort) if save_belongs_to_association(reflection) == false }
             before_save save_method
           end
 
           if reflection.validate? && !method_defined?(validation_method)
             method = (collection ? :validate_collection_association : :validate_single_association)
-            define_non_cyclic_method(validation_method) { send(method, reflection) }
+            define_non_cyclic_method(validation_method) { send(method, reflection) || nil }
```

Comment:
This means we can remove the `|| nil` when we remove the deprecation warning. If that is really the case a TODO comment explaining this would be good.